### PR TITLE
Unify only answers

### PR DESCRIPTION
**Pull request by kasper-piskorski**
- no longer unifying queries
- instead unifying answer objects
- cleanup of the pipeline

### CODE REVIEW

**Comment by kasper-piskorski** [comment]
please retest this please - timeout

**Comment by fppt** [comment]
@kasper-piskorski  rebase needed

**Comment by kasper-piskorski** [comment]
please retest this please - Command execution failed

**Comment by kasper-piskorski** [comment]
anyone wanna review this?

**Comment by fppt** [comment]
I will take a look

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/query/ReasonerAtomicQueryIterator.java [hunk 1]
```diff
@@ -90,30 +99,56 @@
...
+                .unify(uInv);
+
+        Set<VarName> headVars = rule.getHead().getVarNames();
+        Iterator<Answer> baseIterator = rule.getBody().iterator(partialSubPrime, subGoals, cache);
+        return Iterators.transform(
+                baseIterator,
+                a -> {
+                    if (a == null) return null;
+                    else {
+                        return a
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/query/ReasonerAtomicQueryIterator.java]: What does this actually do? I have been staring at it for quite some time and can't figure it out. Maybe a comment stating what this does?

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/rule/InferenceRule.java [hunk 2]
```diff
@@ -151,37 +151,47 @@ public Atom getRuleConclusionAtom() {
...
     /**
-     * @param parentAtom from which constrained are propagated
-     * @return the inference rule with constraints
+     * @param parentAtom atom containing constraints (parent)
+     * @param u rule unifier
+     * @param pu permutation unifier
+     * @return rule with propagated constraints from parent
      */
-    public  InferenceRule propagateConstraints(Atom parentAtom){
+    public InferenceRule propagateConstraints(Atom parentAtom, Unifier u, Unifier pu){
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/rule/InferenceRule.java]: Avoid symbolic programming where please. To an outsider `p` and `pu`don't provide any context. I know you clarify with the javadocs but often the code gets updated without the javadocs being updated so things become even more confusing.

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/query/ReasonerAtomicQuery.java [hunk 3]
```diff
@@ -296,19 +306,31 @@ public Unifier getUnifier(ReasonerQuery p){
...
-    public Stream<Answer> answerStream(Set<ReasonerAtomicQuery> subGoals,
+    Stream<Answer> answerStream(Set<ReasonerAtomicQuery> subGoals,
                                        Cache<ReasonerAtomicQuery, ?> cache,
                                        Cache<ReasonerAtomicQuery, ?> dCache,
                                        boolean materialise,
                                        boolean explanation,
                                        boolean differentialJoin){
         boolean queryAdmissible = !subGoals.contains(this);
 
+        LOG.trace("AQ: " + this);
```
Comment (kasper-piskorski) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/query/ReasonerAtomicQuery.java]: Atomic Query :) This is intentionally a shorthand to make the logs more readable.

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/query/ReasonerAtomicQueryIterator.java (hunk 1 above)
Comment (kasper-piskorski) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/query/ReasonerAtomicQueryIterator.java]: The iterator transform basically unifies the answer from a rule to be an answer to the query.

**Comment by kasper-piskorski** [comment]
The Var to VarPattern refactor is in, Felix is still working on the VarName to Var one.

**Comment by kasper-piskorski** [comment]
rereview please

**Pull request merged**